### PR TITLE
Convert metadata classes to records

### PR DESCRIPTION
**Pull request by wendigo**

### CODE REVIEW

File: core/trino-main/src/main/java/io/trino/execution/scheduler/DynamicSplitPlacementPolicy.java [hunk 1]
```diff
@@ -36,7 +35,7 @@ public DynamicSplitPlacementPolicy(NodeSelector nodeSelector, Supplier<? extends
     }
 
     @Override
-    public SplitPlacementResult computeAssignments(Set<Split> splits)
+    public SplitPlacementResult computeAssignments(List<Split> splits)
```
Comment (ebyhr) [inline core/trino-main/src/main/java/io/trino/execution/scheduler/DynamicSplitPlacementPolicy.java:38]: > Convert Split to record

I'm not sure if we really want to change Split to record causing the collection type change. I would leave to other reviewers.

File: core/trino-main/src/main/java/io/trino/metadata/OutputTableHandle.java [hunk 2]
```diff
@@ -24,48 +22,14 @@
 
 import static java.util.Objects.requireNonNull;
 
-public final class OutputTableHandle
+public record OutputTableHandle(CatalogHandle catalogHandle, SchemaTableName tableName, ConnectorTransactionHandle transactionHandle, ConnectorOutputTableHandle connectorHandle)
```
Comment (ebyhr) [inline core/trino-main/src/main/java/io/trino/metadata/OutputTableHandle.java:25]: Are we excluding  SchemaTableName from hashCode & equals intentionally? Or is it oversight in https://github.com/trinodb/trino/pull/12869?

File: core/trino-main/src/main/java/io/trino/metadata/SignatureBinder.java [hunk 3]
```diff
@@ -1053,12 +1053,10 @@ public enum RelationshipType
...
     }
 
-    private static final class FunctionTypeVariables
+    private record FunctionTypeVariables(FunctionBinding functionBinding)
             implements TypeVariables
     {
-        private final FunctionBinding functionBinding;
-
-        public FunctionTypeVariables(FunctionBinding functionBinding)
+        private FunctionTypeVariables(FunctionBinding functionBinding)
```
Comment (ebyhr) [inline core/trino-main/src/main/java/io/trino/metadata/SignatureBinder.java:1059]: Why not use compact constructor?

File: core/trino-main/src/main/java/io/trino/metadata/TableVersion.java [hunk 4]
```diff
@@ -16,31 +16,4 @@
...
-    {
-        return objectType;
-    }
-
-    public Object getPointer()
-    {
-        return pointer;
-    }
-}
+public record TableVersion(PointerType pointerType, Type objectType, Object pointer) {}
```
Comment (ebyhr) [inline core/trino-main/src/main/java/io/trino/metadata/TableVersion.java:19]: nit: Add compact constructor with requireNonNull. Follow-up is fine.

File: core/trino-main/src/main/java/io/trino/execution/scheduler/DynamicSplitPlacementPolicy.java (hunk 1 above)
Comment (wendigo) [reply core/trino-main/src/main/java/io/trino/execution/scheduler/DynamicSplitPlacementPolicy.java:38]: Records have generated equals and hashCode which is causing issues if you put them into a set. I've discussed that with @losipiuk and @findepi. It seems thar Sets where used just for the sake of tests

Comment (findepi) [reply core/trino-main/src/main/java/io/trino/execution/scheduler/DynamicSplitPlacementPolicy.java:38]: cc @martint @dain

Comment (wendigo) [reply core/trino-main/src/main/java/io/trino/execution/scheduler/DynamicSplitPlacementPolicy.java:38]: I can extract this change to a separate PR

Comment (wendigo) [reply core/trino-main/src/main/java/io/trino/execution/scheduler/DynamicSplitPlacementPolicy.java:38]: Extracted to https://github.com/trinodb/trino/pull/21583. I'll drop this commit here

File: core/trino-main/src/main/java/io/trino/metadata/OutputTableHandle.java (hunk 2 above)
Comment (wendigo) [reply core/trino-main/src/main/java/io/trino/metadata/OutputTableHandle.java:25]: Seems like a omission. Let me remove this equals and hashCode in the separate commit.

**Comment by wendigo** [comment]
I've dropped controversial Split changes. Other commits were mechanical in nature.

File: core/trino-main/src/main/java/io/trino/metadata/IndexHandle.java [hunk 5]
```diff
@@ -13,70 +13,19 @@
...
 import io.trino.spi.connector.CatalogHandle;
 import io.trino.spi.connector.ConnectorIndexHandle;
 import io.trino.spi.connector.ConnectorTransactionHandle;
 
-import java.util.Objects;
-
 import static java.util.Objects.requireNonNull;
 
-public final class IndexHandle
+public record IndexHandle(CatalogHandle catalogHandle, ConnectorTransactionHandle transactionHandle, ConnectorIndexHandle connectorHandle)
```
Comment (findepi) [inline core/trino-main/src/main/java/io/trino/metadata/IndexHandle.java:22]: wrap

File: core/trino-main/src/main/java/io/trino/metadata/InsertTableHandle.java [hunk 6]
```diff
@@ -13,70 +13,19 @@
...
 import io.trino.spi.connector.CatalogHandle;
 import io.trino.spi.connector.ConnectorInsertTableHandle;
 import io.trino.spi.connector.ConnectorTransactionHandle;
 
-import java.util.Objects;
-
 import static java.util.Objects.requireNonNull;
 
-public final class InsertTableHandle
+public record InsertTableHandle(CatalogHandle catalogHandle, ConnectorTransactionHandle transactionHandle, ConnectorInsertTableHandle connectorHandle)
```
Comment (findepi) [inline core/trino-main/src/main/java/io/trino/metadata/InsertTableHandle.java:22]: wrap

File: core/trino-main/src/main/java/io/trino/metadata/OutputTableHandle.java (hunk 2 above)
Comment (findepi) [inline core/trino-main/src/main/java/io/trino/metadata/OutputTableHandle.java:23]: wrap

File: core/trino-main/src/main/java/io/trino/metadata/QualifiedObjectName.java [hunk 7]
```diff
@@ -31,8 +29,7 @@
 import static io.trino.metadata.MetadataUtil.checkObjectName;
 import static java.util.Objects.requireNonNull;
 
-@Immutable
-public class QualifiedObjectName
+public record QualifiedObjectName(String catalogName, String schemaName, String objectName)
```
Comment (findepi) [inline core/trino-main/src/main/java/io/trino/metadata/QualifiedObjectName.java:32]: move to separate PR please

File: core/trino-main/src/main/java/io/trino/metadata/ResolvedFunction.java [hunk 8]
```diff
@@ -27,122 +25,36 @@
...
 
 import java.util.Map;
-import java.util.Objects;
 import java.util.Set;
 
 import static com.google.common.base.Preconditions.checkArgument;
 import static java.util.Objects.requireNonNull;
 
-public class ResolvedFunction
+public record ResolvedFunction(
```
Comment (findepi) [inline core/trino-main/src/main/java/io/trino/metadata/ResolvedFunction.java:33]: move to separate PR please

File: core/trino-main/src/main/java/io/trino/metadata/TableExecuteHandle.java [hunk 9]
```diff
@@ -13,81 +13,30 @@
...
-import java.util.Objects;
-
 import static java.util.Objects.requireNonNull;
 
 /**
  * TableExecuteHandle wraps connectors ConnectorTableExecuteHandle which identifies instance of executing
  * specific table procedure o specific table. See {#link {@link ConnectorTableExecuteHandle}} for more details.
  */
-public final class TableExecuteHandle
+public record TableExecuteHandle(CatalogHandle catalogHandle, ConnectorTransactionHandle transactionHandle, ConnectorTableExecuteHandle connectorHandle)
```
Comment (findepi) [inline core/trino-main/src/main/java/io/trino/metadata/TableExecuteHandle.java:26]: wrap

File: core/trino-main/src/main/java/io/trino/metadata/TableFunctionHandle.java [hunk 10]
```diff
@@ -13,46 +13,18 @@
...
-import com.fasterxml.jackson.annotation.JsonCreator;
-import com.fasterxml.jackson.annotation.JsonProperty;
 import io.trino.spi.connector.CatalogHandle;
 import io.trino.spi.connector.ConnectorTransactionHandle;
 import io.trino.spi.function.table.ConnectorTableFunctionHandle;
 
 import static java.util.Objects.requireNonNull;
 
-public class TableFunctionHandle
+public record TableFunctionHandle(CatalogHandle catalogHandle, ConnectorTableFunctionHandle functionHandle, ConnectorTransactionHandle transactionHandle)
```
Comment (findepi) [inline core/trino-main/src/main/java/io/trino/metadata/TableFunctionHandle.java:22]: wrap

File: core/trino-main/src/main/java/io/trino/metadata/TableHandle.java [hunk 11]
```diff
@@ -13,49 +13,19 @@
...
 import io.trino.spi.connector.CatalogHandle;
 import io.trino.spi.connector.ConnectorTableHandle;
 import io.trino.spi.connector.ConnectorTransactionHandle;
 
-import java.util.Objects;
-
 import static java.util.Objects.requireNonNull;
 
-public final class TableHandle
+public record TableHandle(CatalogHandle catalogHandle, ConnectorTableHandle connectorHandle, ConnectorTransactionHandle transaction)
```
Comment (findepi) [inline core/trino-main/src/main/java/io/trino/metadata/TableHandle.java:22]: wrap. move to separate PR please

File: core/trino-main/src/main/java/io/trino/metadata/TableMetadata.java [hunk 12]
```diff
@@ -22,35 +22,19 @@
...
     {
         return new QualifiedObjectName(catalogName.toString(), metadata.getTable().getSchemaName(), metadata.getTable().getTableName());
     }
 
-    public CatalogName getCatalogName()
-    {
-        return catalogName;
-    }
-
-    public ConnectorTableMetadata getMetadata()
```
Comment (findepi) [inline core/trino-main/src/main/java/io/trino/metadata/TableMetadata.java:49]: when converting TableMetadata to record, other getter-like methods should IMO be converted to, otherwise the class API becomes inconsistent: getQualifiedName, getTable, getColumns


after applying, move to separate PR pelase

File: core/trino-main/src/main/java/io/trino/metadata/TableProperties.java [hunk 13]
```diff
@@ -73,53 +72,12 @@ public Optional<DiscretePredicates> getDiscretePredicates()
         return tableProperties.getDiscretePredicates();
     }
 
-    public static class TablePartitioning
+    public record TablePartitioning(PartitioningHandle partitioningHandle, List<ColumnHandle> partitioningColumns, boolean singleSplitPerPartition)
```
Comment (findepi) [inline core/trino-main/src/main/java/io/trino/metadata/TableProperties.java:75]: wrap

File: core/trino-main/src/main/java/io/trino/metadata/TableSchema.java [hunk 14]
```diff
@@ -23,35 +23,19 @@
...
     }
 
-    public CatalogName getCatalogName()
-    {
-        return catalogName;
-    }
-
-    public ConnectorTableSchema getTableSchema()
-    {
-        return tableSchema;
```
Comment (findepi) [inline core/trino-main/src/main/java/io/trino/metadata/TableSchema.java:52]: same comment as in TableMetadata. keep them together

**Review by findepi — APPROVED** [review]
LGTM %

**Comment by findepi** [comment]
There are related test failures.
Since this PR has many commits, it's impractical to re-review.
it would be best to move out anything already marked for moving out, and also anything problematic (eg causing test failures or needing any non-mechanical changes).

File: core/trino-main/src/main/java/io/trino/metadata/ResolvedFunction.java [hunk 15]
```diff
@@ -27,122 +25,36 @@
...
-        return typeDependencies;
-    }
-
-    @JsonProperty
-    public Set<ResolvedFunction> getFunctionDependencies()
-    {
-        return functionDependencies;
-    }
-
     public CatalogSchemaFunctionName getName()
```
Comment (findepi) [inline core/trino-main/src/main/java/io/trino/metadata/ResolvedFunction.java:55]: ResolvedFunction conversion to record adds "name" into JSON form.
You need explicit `@JsonIgnore` on this. You can link to https://github.com/airlift/airlift/pull/1141 in a TODO comment.

File: core/trino-main/src/main/java/io/trino/metadata/ResolvedFunction.java (hunk 8 above)
Comment (wendigo) [reply core/trino-main/src/main/java/io/trino/metadata/ResolvedFunction.java:34]: https://github.com/trinodb/trino/pull/21593
